Add get_address_proto to handle trailing "tls/ws" in multiaddr

jochasinga opened the pull request:
Before returning `Ws` protocol when hitting the last trailing `/ws` in a multiaddr,
check to see if the next path segment is `/tls`. If it is, then return `Wss` instead.
This addresses #2449 the deprecated `/wss` specifier for the new `/tls/ws` in the spec.

jochasinga committed:
Add get_address_proto to handle trailing "tls/ws" in multiaddr

Before returning Ws protocol when hitting the last trailing `/ws` in a multiaddr,
check to see if the next path segment is `/tls`. If it is, then return Wss instead.
This addresses #2499 the deprecated `/wss` specifier for the new `/tls/ws` in the spec.

jochasinga committed:
Guard against an edge case in which an addr has a trailing "/tls/wss"

This was from the last test which had the typo and happily passed.

thomaseizinger commented on transports/websocket/src/framed.rs:
```diff
@@ -98,6 +98,31 @@ impl<T> WsConfig<T> {
     }
+
+    /// Look backward to if addr has a trailing "/ws" and returns a Wss protocol if "/tls" was found.
+    pub(crate) fn get_address_proto(
```
Does this have to be on `Self`? Could be a free function :)

thomaseizinger replied:
Also, we could refactor this to not mutate its argument and instead consume the argument and return a new address.

jochasinga replied:
Sounds good! Thanks. You mean a public function? Nothing private about it, just feels like it's a helper rather than an API.

jochasinga replied:
And by not mutating its argument and returning a new address, do you mean the function also returns another `Multiaddr` back to the caller? i.e. something like

```rust
fn get_address_proto(addr: Multiaddr) -> Result<(bool, Protocol<'static>, Multiaddr), String>
```

Consuming the `Multiaddr` will likely cause one additional clone as opposed to borrowing mutably since we need to pop the `Multiaddr` one more time to peek at the second to last path segment while keeping the original addr for the caller.

thomaseizinger replied:
> Consuming the Multiaddr will likely cause one additional clone

Looking at the current usage, it is already cloned before being passed in or am I missing something?

The reasoning was: If the caller is okay with mutation (which it currently seems to be), then they can just pass over ownership and use the new return value so it shouldn't make a difference but a non-mutable API would be easier to understand IMO.

thomaseizinger replied:
> You mean a public function? Nothing private about it, just feels like it's a helper rather than an API.

Uhm, still pub(crate) or whatever it needs to be but just not on `Self` and instead free-standing (outside an `impl` block).

jochasinga replied:
How's this? [1e2117a](https://github.com/libp2p/rust-libp2p/pull/2487/commits/1e2117ad5a7046b6123f34e627851d075169d708)

thomaseizinger replied:
Yeah that is much easier to understand IMO, thank you :)

thomaseizinger reviewed (commented):
Thanks!

Some ideas for how this could be improved :)

jochasinga committed:
Refactor get_address_proto as a free-standing function and move multiaddr into it

Previously, the function borrow a mutable multiaddr reference from the caller without
returning it. To make the API simpler, now it's moved into the function and the mutated
multiaddr is returned as part of the triplet tuples.

jochasinga committed:
Refactor to match statement for readability

Refactor get_address_proto into a tuple-matching statement and use
`Iter::last()` to peek into the next protocol segment instead of cloning.

jochasinga committed:
Handle wildcard pattern separately

thomaseizinger commented on transports/websocket/src/framed.rs:
```diff
@@ -142,14 +150,15 @@ where
                 ListenerEvent::NewAddress(mut a) => {
                     a = a.with(proto.clone());
                     debug!("Listening on {}", a);
+                    println!("Listening on {}", a);
```
I don't think we should include `println!`s.

jochasinga replied:
Of course that slipped through the crack.

thomaseizinger reviewed (commented):
I can't comment on what @tomaka said unfortunately. I am not familiar enough with the code here.

jochasinga committed:
Oops, remove print statement
